R/check-up-to-date.R

```diff
@@ -0,0 +1,146 @@
+#' There is no `add_up_to_date()` function because **if you would like to add
+#' these columns to a `bbi_log_df` tibble** you can use [add_config()], which
+#' contains `model_has_changed` and `data_has_changed` columns. Please note:
+#' these contain the opposite boolean values (`check_up_to_date()` returns
+#' `TRUE` if up to date, `*_has_changed` returns `TRUE` if _changed_).
+#'
+#' **The returned value is invisible because a message is printed** alerting the
+#' user of the specific files that have changed, if any. This facilitates
+#' calling the function for this side effect without explicitly handling the
+#' returned value.
+#'
+#' @export
+check_up_to_date <- function(.bbi_object, ...) {
+  UseMethod("check_up_to_date")
+}
+
+#' @export
+check_up_to_date.bbi_nonmem_model <- function(.bbi_object, ...) {
+  check_up_to_date_nonmem(.bbi_object)
+}
+
+#' @export
+check_up_to_date.bbi_nonmem_summary <- function(.bbi_object, ...) {
+  check_up_to_date_nonmem(.bbi_object)
+}
+
+#' @export
+check_up_to_date.bbi_log_df <- function(.bbi_object, ...) {
+
+  check_list <- map(.bbi_object[[ABS_MOD_PATH]], function(.p) {
+    tryCatch(
+      check_up_to_date(read_model(.p)),
+      error = function(.e) {
+        .error_msg <- paste(as.character(.e$message), collapse = " -- ")
+        if (grepl(CHECK_UP_TO_DATE_ERR_MSG, .error_msg, fixed = TRUE)) {
+          message(.error_msg)
+          return(as.logical(c(model = NA, data = NA)))
+        } else {
+          stop(.e)
+        }
+      }
+    )
+  })
+
+  names(check_list) <- .bbi_object[[RUN_ID_COL]]
+  return(invisible(check_list))
+}
+
+####################################
+# PRIVATE implementation functions
+####################################
+
+
+#' Private implementation to check that NONMEM model is up-to-date
+#'
+#' Specifically, check that control stream and data file on disk have not
+#' changed since the model was run. This is accomplished by taking their md5
+#' hashes and comparing it to the hashes stored in `bbi_config.json`.
+#'
+#' @importFrom jsonlite fromJSON
+#' @importFrom fs file_exists
+#' @importFrom tidyr replace_na
+#'
+#' @inheritParams check_up_to_date
+#'
+#' @keywords internal
+check_up_to_date_nonmem <- function(.mod) {
+  config_path <- file.path(get_output_dir(.mod, .check_exists = FALSE), "bbi_config.json")
+  if (!fs::file_exists(config_path)) {
+    stop(paste(glue("Model {get_model_id(.mod)}:"), CHECK_UP_TO_DATE_ERR_MSG))
+  }
+  config <- jsonlite::fromJSON(config_path)
+
+  # check necessary files for changes
+  model_file <- get_model_path(.mod)
+  data_file <- get_data_path(.mod)
+
+  changed_files <- c(
+    config[[CONFIG_MODEL_MD5]] != tools::md5sum(model_file),
+    config[[CONFIG_DATA_MD5]] != tools::md5sum(data_file)
+  )
+
+  any_changes <- any(changed_files)
+  if(isTRUE(any_changes)) {
+    message(paste(
+      glue("The following files have changed in {get_model_id(.mod)}"),
+      paste("*", names(which(changed_files)), collapse = "\n"),
```

Comment:
maybe this is just a me thing but it's kinda weird to mix `paste` and `glue`.. especially because `glue` has the same base functionality as `paste`. I.e. this could be rewritten:

```
message(glue(
glue("The following files have changed in {get_model_id(.mod)}"),
glue_collapse("*", names(which(changed_files)), sep = "\n"),
sep = "\n"
))
```

paste works so this isn't wrong I just don't see why we are using it if we have already imported the glue package

Reply:
yea, I see that. I always just use paste unless I explicitly need the `{}` glue thing. I would change it here but... I do this all over the place so consistency is really not within reach.